File: play-with-lambda-expressions/play_with_lambda/adapters/random_gen.py
"""Random closed-term generator."""

from __future__ import annotations

import random

from ..domain import Abs, App, Term, Var


class RandomTermGenerator:
    """Generates random closed lambda terms via recursive construction.

    Guarantees that every generated term is closed (no free variables)
    by tracking in-scope bindings and forcing abstractions when the
    scope is empty.
    """

    def generate(self, max_depth: int) -> Term:
        """Produce a random closed term with AST depth at most max_depth.

        Args:
            max_depth: Maximum nesting depth of the generated AST.

        Returns:
            A closed Term.
        """
        return self._gen(max_depth, [])

    def _gen(self, depth: int, scope: list[Var]) -> Term:
        """Recursively build a random term.

        Args:
            depth: Remaining depth budget.
            scope: Variables currently in scope (available for Var nodes).

        Returns:
            A Term whose free variables are a subset of scope.
        """
        if not scope:
            # No variables in scope — must introduce one via Abs.
            fresh = self._fresh_param(scope)
            return Abs(fresh, self._gen(depth - 1, scope + [fresh]))

        if depth <= 0:
            # Depth exhausted — return a variable from scope.
            return random.choice(scope)

        # Weighted choice: bias toward Var as depth decreases.
        var_weight = 3
        abs_weight = 2
        app_weight = 2
        roll = random.random() * (var_weight + abs_weight + app_weight)

        if roll < var_weight:
            return random.choice(scope)
        elif roll < var_weight + abs_weight:
            fresh = self._fresh_param(scope)
            return Abs(fresh, self._gen(depth - 1, scope + [fresh]))
        else:
            return App(
                self._gen(depth - 1, scope),
                self._gen(depth - 1, scope),
            )

    @staticmethod
    def _fresh_param(scope: list[Var]) -> Var:
        """Pick a fresh variable name not in scope.

        Args:
            scope: Currently in-scope variables.

        Returns:
            A Var with a name not used in scope.
        """
        used = {v.name for v in scope}
        # Try single letters first, then primed versions.
        for suffix_count in range(10):
            suffix = "'" * suffix_count
            for ch in "abcdefghijklmnopqrstuvwxyz":
                name = ch + suffix
                if name not in used:
                    return Var(name)
        # Fallback — should never be reached in practice.
        return Var(f"v{len(scope)}")

File: play-with-lambda-expressions/src/lambda_calculus/src/adapters/random_gen.rs
//! Random closed-term generator.

use rand::Rng;
use crate::domain::Term;
use crate::ports::TermGenerator;

/// Generates random closed lambda terms via recursive construction.
///
/// Guarantees that every generated term is closed (no free variables)
/// by tracking in-scope bindings and forcing abstractions when the
/// scope is empty.
pub struct RandomTermGenerator {
    rng: rand::rngs::ThreadRng,
}

impl RandomTermGenerator {
    pub fn new() -> Self {
        Self { rng: rand::rng() }
    }

    fn build(&mut self, depth: usize, scope: &[String]) -> Term {
        if scope.is_empty() {
            // No variables in scope — must introduce one via Abs.
            let fresh = Self::fresh_param(scope);
            let mut new_scope = scope.to_vec();
            new_scope.push(fresh.clone());
            let body = self.build(depth.saturating_sub(1), &new_scope);
            return Term::abs(&fresh, body);
        }

        if depth == 0 {
            // Depth exhausted — return a variable from scope.
            let idx = self.rng.random_range(0..scope.len());
            return Term::var(&scope[idx]);
        }

        // Weighted choice: Var(1), Abs(2), App(3)
        let roll: f64 = self.rng.random::<f64>() * 6.0;
        if roll < 1.0 {
            // Var
            let idx = self.rng.random_range(0..scope.len());
            Term::var(&scope[idx])
        } else if roll < 3.0 {
            // Abs
            let fresh = Self::fresh_param(scope);
            let mut new_scope = scope.to_vec();
            new_scope.push(fresh.clone());
            let body = self.build(depth - 1, &new_scope);
            Term::abs(&fresh, body)
        } else {
            // App
            let func = self.build(depth - 1, scope);
            let arg = self.build(depth - 1, scope);
            Term::app(func, arg)
        }
    }

    fn fresh_param(scope: &[String]) -> String {
        let used: std::collections::HashSet<&str> =
            scope.iter().map(|s| s.as_str()).collect();
        for suffix_count in 0..10 {
            let suffix: String = std::iter::repeat('\'').take(suffix_count).collect();
            for ch in 'a'..='z' {
                let name = format!("{}{}", ch, suffix);
                if !used.contains(name.as_str()) {
                    return name;
                }
            }
        }
        format!("v{}", scope.len())
    }
}

impl TermGenerator for RandomTermGenerator {
    fn generate(&mut self, max_depth: usize) -> Term {
        self.build(max_depth, &[])
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::domain::is_closed;

    #[test]
    fn generated_terms_are_closed() {
        let mut generator = RandomTermGenerator::new();
        for depth in 1..=5 {
            for _ in 0..20 {
                let term = generator.generate(depth);
                assert!(is_closed(&term), "generated term is not closed: {}", term);
            }
        }
    }

    #[test]
    fn fresh_param_avoids_used() {
        let scope = vec!["a".to_string(), "b".to_string()];
        let fresh = RandomTermGenerator::fresh_param(&scope);
        assert!(!scope.contains(&fresh));
    }

    #[test]
    fn fresh_param_empty_scope() {
        let fresh = RandomTermGenerator::fresh_param(&[]);
        assert_eq!(fresh, "a");
    }
}
